Docs page on interoperability (#7992)
pydata/xarray: doc/internals/internal-design.rst

    and to merge the `xarray-datatree <https://github.com/xarray-contrib/datatree>`_ package into xarray's main repository.

Internally private :ref:`lazy indexing classes <internal design.lazy indexing>` are used to avoid loading more data than necessary,
and flexible indexes classes (derived from :py:class:`~xarray.indexes.Index`) provide performant label-based lookups.

.. _internal design.data structures:

Data Structures
---------------

The :ref:`data structures` page in the user guide explains the basics and concentrates on user-facing behavior,
whereas this section explains how xarray's data structure classes actually work internally.

.. _internal design.data structures.variable:

Variable Objects
~~~~~~~~~~~~~~~~

The core internal data structure in xarray is the :py:class:`~xarray.Variable`,
which is used as the basic building block behind xarray's
:py:class:`~xarray.Dataset`, :py:class:`~xarray.DataArray` types. A
:py:class:`~xarray.Variable` consists of:

- ``dims``: A tuple of dimension names.
- ``data``: The N-dimensional array (typically a NumPy or Dask array) storing
  the Variable's data. It must have the same number of dimensions as the length
  of ``dims``.

> - ``attrs``: An ordered dictionary of metadata associated with this array. By
>   convention, xarray's built-in operations never use this metadata.
> - ``encoding``: Another ordered dictionary used to store information about how

Reviewer: ```suggestion
- ``attrs``: A dictionary of metadata associated with this array. By
  convention, xarray's built-in operations never use this metadata.
- ``encoding``: Another dictionary used to store information about how
```
If I recall correctly attrs and encoding are simple dicts, we don't do anything fancy with them.

Not important but an example where we use `attrs` directly is to add units to plots with `attrs["units"]`.

Author: I think you're right. (I actually didn't write this section, just moved it, so I wonder if it changed at some point...)
